File: hardware/grid.py
import util

class Grid:
    def __init__(self, bounding_box, subdivisions):
        aspect_ratio = bounding_box.get_aspect_ratio()
        util.debug(f'grid: top_left={bounding_box.top_left} bottom_right={bounding_box.bottom_right} subdivisions={subdivisions}')
        self.bounding_box = bounding_box
        self.subdivisions = subdivisions
        self.table = {}

    def add_segment(self, segment, index):
        if index in self.table:
            list = self.table[index]
        else:
            list = []
            self.table[index] = list

        list.append(segment)

    def get_index(self, lat, lon):
        ### lat == y, lon == x?
        latfrac = self.bounding_box.get_lat_fraction(lat, False)
        if latfrac is None:
            return -1
        lonfrac = self.bounding_box.get_long_fraction(lon, False)
        if lonfrac is None:
            return -1
        index = (self.subdivisions * int(round(self.subdivisions * latfrac))
            + int(round(self.subdivisions * lonfrac)))
        return index

    def get_segment_list(self, lat, lon):
        index = self.get_index(lat, lon)

        if index in self.table:
            return self.table[index]
        else:
            return None

    def __str__(self):
        return '{\n  bounding_box: ' + str(self.bounding_box) + '\n  subdivisions: ' + str(self.subdivisions) + '\n  len(table): ' + str(len(self.table)) + '\n}'

File: server/app-engine/static/grid.js
class Grid {
    constructor(boundingBox, subdivisions){
        this.boundingBox = boundingBox;
        this.aspectRatio = boundingBox.getAspectRatio();
        this.subdivisions = subdivisions;
        this.table = {};
        util.log(`grid: top_left=${JSON.stringify(boundingBox.topLeft)} bottom_right=${JSON.stringify(boundingBox.bottomRight)} subdivisions=${subdivisions}`)
    }

    addSegment(segment, index){
        // util.log("addSegment()");
        let list = null;
        if (index in this.table){
            list = this.table[index];
        }
        else{
            list = [];
            this.table[index] = list;
        }
        list.push(segment);
    }

    getIndex(lat, lon){
        let latFrac = this.boundingBox.getLatFraction(lat, false)
        if (latFrac === null){
            return -1;
        }
        let lonFrac = this.boundingBox.getLongFraction(lon, false)
        if (lonFrac === null){
            return -1;
        }
        let index = (this.subdivisions * parseInt(Math.round(this.subdivisions * latFrac))
            + parseInt(Math.round(this.subdivisions * lonFrac)))
        return index;
    }

    getSegmentList(lat, lon){
        util.log("getSegmentList()");
        let index = this.getIndex(lat, lon);
        if (index in this.table){
            return this.table[index];
        } else return null;
    }
}